server/src/test/java/org/elasticsearch/cluster/NodeConnectionsServiceTests.java

```diff
@@ -77,65 +87,212 @@
         return nodes;
     }
 
-    private ClusterState clusterStateFromNodes(List<DiscoveryNode> nodes) {
+    private DiscoveryNodes discoveryNodesFromList(List<DiscoveryNode> discoveryNodes) {
         final DiscoveryNodes.Builder builder = DiscoveryNodes.builder();
-        for (DiscoveryNode node : nodes) {
-            builder.add(node);
+        for (final DiscoveryNode discoveryNode : discoveryNodes) {
+            builder.add(discoveryNode);
         }
-        return ClusterState.builder(new ClusterName("test")).nodes(builder).build();
+        return builder.build();
     }
 
-    public void testConnectAndDisconnect() {
-        List<DiscoveryNode> nodes = generateNodes();
-        NodeConnectionsService service = new NodeConnectionsService(Settings.EMPTY, threadPool, transportService);
-
-        ClusterState current = clusterStateFromNodes(Collections.emptyList());
-        ClusterChangedEvent event = new ClusterChangedEvent("test", clusterStateFromNodes(randomSubsetOf(nodes)), current);
-
-        service.connectToNodes(event.state().nodes());
-        assertConnected(event.state().nodes());
+    public void testConnectAndDisconnect() throws Exception {
+        final NodeConnectionsService service = new NodeConnectionsService(Settings.EMPTY, threadPool, transportService);
 
-        service.disconnectFromNodesExcept(event.state().nodes());
-        assertConnectedExactlyToNodes(event.state());
+        final AtomicBoolean stopReconnecting = new AtomicBoolean();
+        final Thread reconnectionThread = new Thread(() -> {
```

Comment:
This thread is never started?

Reply:
Oops :) 0de7438f13b62df1d653ca63b2d91ba5b858b832